TowerScheduler/scheduler.py

```diff
@@ -0,0 +1,442 @@
+            time.sleep(1) # yield
+
+            # Python sleep, sleep timer is offline
+            self.logger.error("Sleep timer failed to shut tower down. " + \
+                "Calling Python's time.sleep for %i seconds", to_sleep)
+            time.sleep(to_sleep)
+
+        else:
+            # if not enough time, call Python sleep for full time
+            self.logger.info("Calling Python's time.sleep for %i seconds",
+                    available_sleep_time)
+            time.sleep(available_sleep_time)
+
+        self.logger.info("Changing state to CheckTime")
+        state_machine.curr_state = CheckTime.get_singleton()
+
+
+class StateMachine:
+    '''
+    The StateMachine class holds the data that needs to be accessed
+    from multiple states and runs an infinte loop of process and update
+    '''
+
+    def __init__( self,
+                ens_list: List[Ensemble],
+                sleep_func: Callable[[int], None] ):
+
+        self.config = Configuration.get_singleton(Path('schedulerConfig.ini'))
+        self.logger = get_logger("State Machine", self.config.log_level)
+
+        # State machine's data on ensemble schedule: list and next to execute
+        self.ens_list: List[Ensemble] = ens_list
+        self.ens_index: int = 0
+
+        # Sleep timer's sleep func to call in order to shut tower down
+        self.sleep_func: Callable[[int], None] = sleep_func
+
+        # Data used when transitioning from end of one day to beginning of next
+        self.day_of_ens: int = 0
+        self.daily_reset: boolean = False
+
+        # Start from wakeup state
+        self.curr_state: State = WakeUp.get_singleton()
+
+    def run_machine(self):
+        while True:
+            self.curr_state.process(self)
+            self.curr_state.update(self)
+
+    @property
+    def wakeup_time(self):
+        return self.config.wakeup_time
+
+    @wakeup_time.setter
+    def wakeup_time(self, sec: int):
+        if sec < self.config.wakeup_time:
+            self.logger.info("New wakeup time %i is less than previous!", sec)
+        self.config.wakeup_time = sec
+
+    @property
+    def shutdown_time(self):
+        return self.config.shutdown_time
+
+    @shutdown_time.setter
+    def shutdown_time(self, sec: int):
+        if sec < self.config.shutdown_time:
+            self.logger.info("New shutdown time %i is less than previous!", sec)
+        self.config.shutdown_time = sec
+
+
+def main(ens_file: Path = "active_ensembles.json"):
+    try:
+        ens_list = Ensemble.list_from_json(ens_file)
+    except FileNotFoundError:
+        logging.exception("Active ensembles file not found. " + \
+                        "Unable to continue.\n")
+        sys.exit()
+    except KeyError:
+        logging.exception("Active ensembles file is improperly formatted. " + \
+                        "Unable to continue.\n")
+        sys.exit()
+
+    control_flow = StateMachine(ens_list, time.sleep)
+    control_flow.run_machine()
```

Comment:
Might not be a terrible idea to just do `StateMachine(ens_list, time.sleep).run_machine()`